spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,120 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.actions;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.datasketches.theta.Sketch;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.puffin.Blob;
+import org.apache.iceberg.puffin.StandardBlobTypes;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import scala.Tuple2;
+
+public class NDVSketchGenerator {
+
+  private NDVSketchGenerator() {}
+
+  public static final String APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY = "ndv";
+  private static final String SNAPSHOT_ID = "snapshot-id";
+
+  static List<Blob> generateNDVSketchesAndBlobs(
+      SparkSession spark, Table table, long snapshotId, Set<String> columnsToBeAnalyzed) {
+    Iterator<Tuple2<String, ThetaSketchJavaSerializable>> columnToNdvSketchPair =
+        computeNDVSketches(spark, table, snapshotId, columnsToBeAnalyzed);
+    Map<String, ThetaSketchJavaSerializable> sketchMap = Maps.newHashMap();
+    columnToNdvSketchPair.forEachRemaining(tuple -> sketchMap.put(tuple._1, tuple._2));
+    return generateBlobs(table, columnsToBeAnalyzed, sketchMap, snapshotId);
+  }
+
+  private static List<Blob> generateBlobs(
+      Table table,
+      Set<String> columns,
+      Map<String, ThetaSketchJavaSerializable> sketchMap,
+      long snapshotId) {
+    return columns.stream()
+        .map(
+            columnName -> {
+              Sketch sketch = sketchMap.get(columnName).getSketch();
+              long ndv = (long) sketch.getEstimate();
+              return new Blob(
+                  StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                  ImmutableList.of(table.schema().findField(columnName).fieldId()),
+                  snapshotId,
+                  table.snapshot(snapshotId).sequenceNumber(),
+                  ByteBuffer.wrap(sketch.toByteArray()),
+                  null,
+                  ImmutableMap.of(APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY, String.valueOf(ndv)));
+            })
+        .collect(Collectors.toList());
+  }
+
+  static Iterator<Tuple2<String, ThetaSketchJavaSerializable>> computeNDVSketches(
```

Comment:
* any reason  this method is not private?
* how about this method return the Map directly to let the main method be cleaner?